### ISSUE TITLE
How to configure emacs-jupyter if jupyter is not global?

### ISSUE THREAD

**Issue by nanounanue** [nnicandro/emacs-jupyter#353]
Hi

I am currently using `direnv` for managing my environments. In particular, I don't have `jupyter` installed globally, it is installed (using  `GUIX` + `direnv`) in the directory that is needed (and only activated when I am visiting that directory).

All the other packages that use `python` or `jupyter` are able to use the local installation, but `emacs-jupyter` fails at startup because the line in `(jupyter    . t)` in the `org-babel-load-languages`. The error thrown is related to the fact that `jupyter` is not in the `path` at startup.

Is it possible to defer that line until needed? Maybe using `.dir-locals.el`? Or what is the recommended approach?

Thanks in advance

**Comment by xor-xor** [comment]
@nanounanue not sure if this is *the* recommended approach, but in my case (I use virtualenvs instead of global jupyter), creating dedicated kernelspecs for the kernels that I want to use solved my problem (see [here](https://github.com/nnicandro/emacs-jupyter/issues/19#issuecomment-530937674)).

Since then, I've written some code in my `~/.emacs` to facilitate this process (adding/removing kernelspecs):
```elisp
(require 'cl-lib)

(defun my--get-venv-names-for-completion ()
  (list (completing-read "venv-name: "
			 (directory-files (expand-file-name "~/envs")
					  nil
					  "^[^._]")
			 nil
			 t)))

(defvar my-jupyter-kernels-dir
  (expand-file-name "~/envs/jupyter_only/share/jupyter/kernels")
  "Location of kernels that can be run via `jupyter-run-repl'")

(defun my-jupyter-add-kernel (kernel)
  "Create a kernelspec for KERNEL and add it for use with `emacs-jupyter'.

The list of kernels available for adding is constructed from
virtual environments found in `~/envs/`."
  (interactive (my--get-venv-names-for-completion))
  (when (member kernel (cl-map 'list 'car (jupyter-available-kernelspecs)))
    (error (format "Kernel \"%s\" is already added" kernel)))
  (let* ((argv (vector (expand-file-name (format "~/envs/%s/bin/python3" kernel))
		       "-m"
		       "ipykernel_launcher"
		       "-f"
		       "{connection_file}"))
	 (content (json-serialize
		   `(:argv ,argv :display_name ,kernel :language "python")))
	 (kernel-dir (format "%s/%s" my-jupyter-kernels-dir kernel)))
    (make-directory kernel-dir)
    (let ((save-silently t))
      (append-to-file content nil (format "%s/kernel.json" kernel-dir))))
  (jupyter-available-kernelspecs t))

(defun my-jupyter-delete-kernel (kernel)
  "Remove KERNEL from available kernelspecs and delete its files."
  (interactive (list (completing-read "kernel to delete: "
				      (cl-map 'list 'car (jupyter-available-kernelspecs))
				      nil
				      t)))
  (when (equal kernel "") (error "No kernel selected"))
  (delete-directory (format "%s/%s" my-jupyter-kernels-dir kernel) t)
  (jupyter-available-kernelspecs t))
```
There are some hard-coded paths specific to my setup, namely `~/envs`, which is a directory where I keep my virtualenvs, and `jupyter_only`, which is a virtualenv with only jupyter package installed in it - it serves as an alternative to global jupyter (I have `$HOME/envs/jupyter_only/bin` appended to my `$PATH`) and kernelspecs are added in there.

**Comment by gmoutso** [comment]
I also use kernels in virtual environments. But I also had the same issue with loading the package. My solution was to install jupyter once _globally_ (using anaconda inside ~/anaconda3/bin/).

**Comment by gmoutso** [comment]
As a comment. I run one server from one conda environment that can access all other kernels in other environments using anaconda's package nb_conda_kernels. But regardless I also have a global jupyter command found in my exec-path.

**Comment by nanounanue** [comment]
Thank you @xor-xor for your advice and code. One question, How do you manage adding jupyter to the `org-babel-load-languages` list? if I add it (using `(jupyter . t)`) I get an error at the startup ...

**Comment by nanounanue** [comment]
Thank you @gmoutso that was my previous setup, but at the moment, I am running `guix` in a foreign distro. If I load `python` at my environment when I log in, it fails (the OS get confused about using the `python` from the OS or the one from `guix`)

So, I delegate `python` to use only in development environments (managed by `direnv`)

**Comment by xor-xor** [comment]
> How do you manage adding jupyter to the org-babel-load-languages list?

@nanounanue I don't think that this requires any special handling, apart from that `(jupyter . t)` has to be the last item on this list. Here's the relevant part from my config:

```elisp
(with-eval-after-load 'org
  (org-babel-do-load-languages 'org-babel-load-languages
                               '((emacs-lisp . t)
				 (python . t)
				 (shell . t)
				 (latex . t)
				 (sql . t)
				 (sqlite . t)
				 (restclient . t)
				 (calc . t)
				 (jupyter . t))))
```
There are some other `setq`s and customizations in this block which I've omitted here as irrelevant, but this should work on its own (given that you have support for all those languages). In fact, yesterday I've been [testing something](https://github.com/nnicandro/emacs-jupyter/issues/357#issue-1050214406) only with this:
```elisp
(with-eval-after-load 'org
  (require 'org-tempo)
  (org-babel-do-load-languages 'org-babel-load-languages
                               '((python . t)
				 (jupyter . t)))
  (add-to-list 'org-structure-template-alist '("jp" . "src jupyter-python")))
```
...and it worked fine (as far as your problem is concerned). And btw, you don't have to wrap this in `with-eval-after-load` - in the past, I used to have its body as top-level entities in my config, and it worked w/o problems.

**Comment by xor-xor** [comment]
One thing that maybe is not obvious here, but seems to be a source of confusion for people used to solutions like virtualenv (including me >2 yrs ago) is that emacs-jupyter is totally unaware of them - it uses what it gets from `(jupyter-locate-python)` (`"/Users/xor-xor/envs/jupyter_only/bin/python3"` in my case), but the way you configure this path is totally up to you (e.g. by manipulating `$PATH` outside of Emacs or using some other mechanisms) - if `jupyter-locate-python` cannot pick it up, then your problem more than likely lies outside of emacs-jupyter.

This python binary returned by the aforementioned command also needs jupyter package somewhere in its `sys.path` - therefore global installation of jupyter works here, but also the approach that I've presented in my previous comments (i.e. to have some dedicated virtualenv with jupyter and its bin dir added to `$PATH`), and probably some other approaches too.

One may wonder: "OK, but the whole point of virtualenv-like solutions is to be able to run jupyter sessions from different dirs/env, with different packages installed etc. So how this global or global-like approach fits in here?" - and here come kernelspecs, which allow you to use paths to _any other virtualenv you want_ (given that jupyter/ipykernel_launcher is installed in there) - you just need to modify this first item of `"argv"` list in `kernel.json` file (changing `"display_name"` would be a good idea too) to some other python, in some other virtualenv, e.g.:
```json
{
 "argv": [
  "python",
  "-m",
  "ipykernel_launcher",
  "-f",
  "{connection_file}"
 ],
 "display_name": "Python 3 (ipykernel)",
 "language": "python",
 "metadata": {
  "debugger": true
 }
```
...is changed to:

```json
{
 "argv": [
  "/Users/xor-xor/envs/some_newly_created_env/python",
  "-m",
  "ipykernel_launcher",
  "-f",
  "{connection_file}"
 ],
 "display_name": "my new env",
 "language": "python",
 "metadata": {
  "debugger": true
 }
```
This way, the aforementioned global (or global-like) jupyter serves only as a dispatcher of sorts - it just launches kernels from the paths specified in its kernelspecs, that's all. If you need to install some packages for one of your kernels, you install them in respective virtualenvs, not into this "global" location.

**Comment by yitang** [comment]
I use the same approach as @nergooi. it would be great we can start emacs without installing Jupyter.

**Comment by akirakyle** [comment]
It should be possible with a similar idea but with an org babel elisp block like so

```
#+begin_src elisp :results silent
(org-babel-do-load-languages
   'org-babel-load-languages
   '(
     (emacs-lisp . t)
     (org . t)
     (shell . t)
     (latex . t)
     (jupyter . t)
     ))
(org-babel-jupyter-aliases-from-kernelspecs t)
#+end_src
```
